### ISSUE TITLE
[RFC] Functional analyzers

### ISSUE THREAD

**Issue by MartijnHols** [WoWAnalyzer/WoWAnalyzer#4541]
> We did some brainstorming on this on Discord and I hope this RFC will flesh out the idea, get more people behind it and serve as a good design document in case the implementation takes a while.

Glossary

- **Functional**: [This article](https://medium.com/javascript-scene/master-the-javascript-interview-what-is-functional-programming-7f218c68b3a0) does a good job explaining it. In short it can be summarized to be pure functions while avoiding state*.
- **Pure functions**: a function which given the same input always returns the same output and has no side-effect. `(a) => a * 5` is pure, `(a) => a + this.previousValue` is not and neither is `(a) => { window.something = 'something'; return a * 5; }`. When passing objects, you should treat them as immutable and return a new instance. e.g. `(a) => { return { ...a, prop: 1 } }` and not `(a) => { a.prop = 1; return a }`.

*Avoiding state
> Let's recap: state is where you have a variable that remembers the last value of something. In a log this might be the Haste level of someone at the current time, which can be accessed today through `this.haste.current` given a dependency on the `Haste` module.

The main reason for the event parsing setup currently implemented in WoWAnalyzer is that it matches what a player would experience in game. If you're at 2:00 in a boss fight your Haste could be 10%, boss at 88% health, you are affected by the Bloodlust buff, etc. Current we let you read this state through various methods, such as `this.haste.current` for the state. This works pretty well for matching what someone would experience in-game but isn't great for code quality and maintenance.

## What is a functional analyzer?

A fully functional analyzer is a function that **takes events** and **outputs a value** without the usage of any dependencies or mutating any state. The events are passed as an argument; `events: Event[]` which can be reduced; `events.reduce(x)`.

Writing a functional analyzer to sum the total damage of a certain ability would be easy:

```ts
const totalFireballDamage = (events: Event[]) => events.reduce((sum, event) => {
  if (event.type === DamageEventType && event.ability.guid === FIREBALL_ID) {
    sum += event.amount
  }

  return sum
})
```

> We would make composers available for commonly used analysis so you won't have to write this code yourself every time but can instead do `const totalFireballDamage = totalDamage(FIREBALL_ID)`.

This matches the "pure function" definition earlier: given the same events it will always output the same damage done. It does not get affected by the state of anything.

Often times you need more complicated analysis with something another module provides such as Haste. To eliminate state, we would need to expand these modules to also fabricate HasteChange events and the likes. This is a great solution to bridge the gap between current analyzers and fully functional analyzers (if we choose to go in that direction) as all current code can continue to work properly.

If we move the Haste state to events you can track Haste by using a variable in your function scope:

```ts
const totalFireballHasteDamage = (events: Event[]) => {
  let haste = 0
  return events.reduce((sum, event) => {
    if (event.type === HasteChangeEventType) {
      haste = event.haste
    }
    if (event.type === DamageEventType) {
      sum += event.amount * haste
    }
    return sum
  }, 0)
}
```

or by looping through the previous events:

```ts
const totalFireballHasteDamage = (events: Event[]) =>
  events.reduce((sum, event, index) => {
    if (event.type === DamageEventType) {
      const haste = findReverse(events, index, (event) => event.type === HasteChangeEventType).haste
      sum += event.amount * haste
    }
    return sum
  }, 0)
```

or we could provide Haste as a property on all other events:

```ts
const totalFireballHasteDamage = (events: Event[]) =>
  events.reduce((sum, event) => {
    if (event.type === DamageEventType) {
      sum += event.amount * event.haste
    }
    return sum
  }, 0)
```

Which solution of these three options would be best will become clear after some experimentation. The important thing to note is that there are various solutions which can result in clean code, which option we choose is not significant to this RFC.

## What are the advantages?

Testing these analyzers no longer requires integration-like tests with test CombatLogParsers and setting up complicated dependencies that make your tests fragile. A functional analyzer test can be as simple as `expect(totalFireballHasteDamage([hasteChangeEvent, castEvent])).toBe(50)`. Both events are simple event objects. That's it.

Functional programming also reduces the complexity of modules mixing state. Many modules have several dependencies. If one module is changed, another can break. This can make it very hard to change or refactor modules. By eliminating dependencies, each analyzer becomes fully standalone. The only dependency is the event object structure, which can be considered a public contract (and with TypeScript this gets easy to typecheck).

Since functional analyzers do not rely on state they can be executed in parallel. This can speed up analysis and makes server-side analysis more viable.

Functional analyzers are pure JavaScript mathematical code, no React or anything mixed in. This reduces the learning curve and makes it a lot easier to execute in Node.JS environments (servers rather than browsers).

## Hold up-- How do we output statistics, suggestions etc?

The interface would be separated from the log analysis. Log analysis outputs all the data that is available, a developer then uses this data to write the interface. This gives more freedom in structuring the interface as ideally for users as possible. It also gives full freedom in structuring the analysis code without having to consider combining stats into a single Statistic.

Imagine statistics is a canvas that you now get full control over.

```ts
const MyStatistics = (combatant, events: Event[]) => {
  <Statistics>
    {combatant.hasTalent(MY_TALENT_ID) && (
      <Statistic size="small">
        <BoringSpellValue
          spell={SPELLS.LIGHT_OF_DAWN_CAST}
          value={playersHitPerCast(events).toFixed(2)}
          label={
            <Trans id="paladin.holy.modules.lightOfDawn.avgHitPerCast">
              Average targets hit per cast
            </Trans>
          }
          className="light-of-dawn-hits-per-cast"
          extra={<PlayerHits performance={performance} />}
        />
      </Statistic>
    )}

    <SomeOtherStat />

    <SomeOtherStat />

    <SomeOtherStat />
  </Statistics>
}
```

You get full control over positioning, layout, etc. Want to add a flexbox container to always put a big panel to the right of the rest of the statistics? Go ahead!

This puts all of the interface code together, gives full control to the developer and keeps the interface separate from the analysis.

## Afterthoughts:

- **active/inactive modules**: this is no longer necessary as the `combatant.hasTalent(MY_TALENT_ID)` check in the interface already takes care of it. If the statement returns false, the functional analyzer is never executed.
- **selectedCombatant**: this could be passed as a second argument so analyzers become `function analyze(events: Event[], info: { combatant: Combatant })` where the `info` arg can get any static data we might want to share (e.g. boss id as well).
- **migrating the codebase**: this is a WIP, I first would like to hear your opinions on the solution. We'll have to deem if it is worth it and then I'm sure we can figure out a clean way to migrate.
- **issues**: similar to statistics, this becomes its own function with similar conditional statements. Instead of outputting a React component, it would output an array with issues though.
- **checklist**: pretty much the same solution as statistics (it's already a lot like this). It would just stop being a module and get a special place.
- **performance**: I checked a few measurements and `reduce` seems *faster* than `for`-loops, but I am unsure about the accuracy of these measurements. As for looping over the events multiple times: given we can parallelize everything I think in the end it should be quicker. We could also make statistics do their analysis lazily to show the results page instantly (but without values) and have the values pop in one-by-one as analysis finishes up.
- We could make **composers** for a lot of things. Total damage done, spell casts, and specs could make their own like beacon healing done.

## New analysis steps

With this solution the analysis steps would change:

0. WCL Events
1. Normalizers (fix inconsistencies)
2. Enhancers (fabricate data that requires complicated analysis e.g. add Haste change events)
3. Analyze
4. Output data in the interface

Each step is clearly separated from the previous steps and has clear in and output.

0. In: report id, out: events array
1. in: events array, out: modified events array
2. in: events array, out: expanded events array
3. in: events array, out: single values
4. in: single values, out: interface

**Comment by MartijnHols** [comment]
First stat and functional suggestion can be found here: https://github.com/WoWAnalyzer/WoWAnalyzer/pull/4556

Not much had to be updated to support this.

**Comment by MartijnHols** [comment]
Renamed stats to metrics.

Functional statistic implementation: https://github.com/WoWAnalyzer/WoWAnalyzer/pull/4560 and https://github.com/WoWAnalyzer/WoWAnalyzer/pull/4562

I think this implementation shows the entire concept.

**Comment by emallson** [comment]
After having used it extensively for the APL checker and talked with people about it, I like using it but am not looking forward to the immense amount of work that transitioning fully to it would require and doubly uninterested in maintaining two separate ways of doing analysis.

However, not all is lost. TS supports anonymous classes! So what we could do is actually write a `toAnalyzer(FunctionalAnalyzer f, EventFilter[] filters = [], Analyzer[] dependencies = []): ? extends Analyzer` method that unifies the internal implementation (basically: no more functional analyzers as such) while still allowing us to write analyzers as functions.

Basically, this is a way for us to get the functional syntax without needing to re-engineer the `CombatLogParser` internals and migrate large amounts of existing class Analyzers to new functional analyzers (or live with two separate styles of analyzers).

The alternative would be writing a `toFunctional(Analyzer a): FunctionalAnalyzer` method that is used by the `CombatLogParser` to map everything into a functional analyzer, but given the amount of work that fully moving to functional analyzers would require, I think that sticking with the class approach will be more practical (and, in the long term, if people largely adopt functional analyzers then we have a significantly easier transition path to fully functional internals at that time).

I do still think that removing the ability for the `Analyzer` class to dispatch events (moving that to the `Normalizer`/`Enhancer` stage) is a good plan for the long term. Much :spaghetti: has been caused by the ability for analyzers to modify anything at any time.

**Comment by mfield** [comment]
I'm kind of new here, maybe this is a weird or half-baked idea. The project already has react and redux, functional analyzers look like reducers. Specs could be organized into stores, with their `CombatLogParser` set up to route events to the right reducers. View layer reads analysis outputs from redux stores, completely isolated from analysis logic.

Anyway, just a thought because functional analyzers reminded me of redux reducers.

**Comment by emallson** [comment]
They effectively are reducers limited to a specific store context (events). We could make interaction with redux a first-class part of it but for the purposes of separating concerns I think limiting that interaction is probably for the best (to avoid similar issues as have been introduced by Analyzers modifying the event stream willy-nilly)

**Comment by MartijnHols** [comment]
I appreciate the comparison with Redux, good observation. One big problem I have with Redux is that it requires you to put all reducers in a nested reducer to make one giant object of data. This concept from the start sounds like spaghetti. The alternative I proposed are just functions. You call them in your React component or don’t, just like you would a regular function. They’re nothing special and involve no magic apart from memoization. 

I think converting functional analyzers to classes wouldn’t really work therefore. They behave completely different to the old analyzers and have no event they could be tied to. They’re rendered at the end of (the classic) analysis and take care of the rest themselves. In other words how it is structured today iirc already has nearly zero code insertion within the CLP event parsing behavior, and so maintenance should be minimal. The only issue would be using functional analysis from classic analyzers, which I argue should be unnecessary as I think new analyzers would be best written functionally.

**Comment by MartijnHols** [comment]
I should say I think someone would need to become ambassador of functional components and extend it with new common ones, extend the base as required as well as build more example statistics with it. If no one is interested in taking on that project, it may be better to scrap the concept entirely as leaving it to die slowly while maintaining two different paradigms would be worse.

**Comment by emallson** [comment]
I can be that ambassador, because I want to continue using them. I just don't want to (a) have two different codepaths for functional vs nonfunctional or (b) embark on a large refactor of the core analysis code (especially since we haven't figured out things like dependencies for functional components).

Additionally, I don't want to require all of the maintainers to re-learn how to write analyzers again unless I'm absolutely confident that we're going to remain with this system for an extended period w/o major changes---and with dependencies etc undetermined and definitely untested I don't have that confidence.

But I used the functional stuff for the APL checker and was very happy with it so I intend to continue using it. :crossed_fingers: we end up with a gradual migration to it.

**Comment by MartijnHols** [comment]
Cool, that's really nice to hear :D

As for the two different codepaths, isn't [this](https://github.com/WoWAnalyzer/WoWAnalyzer/blob/shadowlands/src/parser/core/CombatLogParser.tsx#L700-L742) all the code required to support functional analyzers right now? The simplicity of it is part of what makes me so excited about it, as that makes it so much easier to maintain (for contrast: I don't think anyone would dare to make any changes in the dependency resolution due to its complexity).

Functional components as I envisioned them do not need dependencies (another aspect making them simpler). You have the different event parsing phases as laid out in the last section of the RFC, by the time the functional analyzers run, all dependencies will already have ran. Things like beacon healing should be normalized in (and until then it happens in the "enhancers" step). What dependency could not be done in these two steps? In any case I did address this on Discord and I think in my implementation (but I never updated the RFC); you just call other metrics from the metrics that you need. There's no need for special or extra code for dependencies - it works just like plain old JavaScript (with memoization on top).

A big part of the migrations would be to turn core modules (i.e. the dependencies) into normalizers. As you said this was already desirable for class(ic) analyzers. I do think normalizers are more complicated to write than regular analyzers - maybe we could provide utilities or new coding patterns to simplify that.

**Comment by emallson** [comment]
I posted an implementation of functional analyzers in terms of class analyzers in the #maintainers channel on discord (link: https://github.com/emallson/WoWAnalyzer/blob/standardize-simplify/src/parser/core/Analyzer.ts#L70-L133).

You are correct about dependencies, I think. Inter-analyzer dependencies are solved by memoization (i.e. importing & calling the analyzer you want the results from). The rest of everything moves to Normalizers/Enhancers, and then the Checklist...honestly needs a lot of ergonomic work anyway so that wouldn't hurt.

I guess let me re-frame my thoughts. This isn't about removing functional analyzers per se, but rather: how do we get there? There are two major concerns:

1. the actual act of refactoring the large existing codebase
2. informing contributors of the update and helping them transition

both of these are more realistically done slowly over time than all in one go during the 10.0 push.

With that in mind, my thinking is that we proceed with something like this:

- [ ] Remove explicit functional analyzer support from the core `CombatLogParser`, replacing it with functions that generate `Analyzer` classes.
- [ ] Begin migration of all `EventDispatch` uses to Normalizers, adding QoL tools as we go to make it easier to write them.
  - [ ] Disallow merging new analyzers that use `EventDispatch`.
- [ ] Update documentation to reflect (default to?) functional analyzers. I think that for most new contributors they'll be easier to write anyway so we'll likely naturally transition to them over time.
- [ ] *Eventually:* switch `CombatLogParser` to understand functional analyzers by default and add functions to generate functional analyzers from class analyzers (which is actually *super* easy to do, not that the converse is particularly difficult either)

This allows us to move in the direction of functional analyzers without any big breaks. It also allows us to back out of it if it turns out that functional analyzers have some problems / don't have broader uptake by contributors.

There are a couple of codebase-level changes I want to do for 10.0 (like the `EventDispatch` one, and adding spelldata via ids rather than manually-curated objects) that will already involve a large chunk of work.

**Comment by emallson** [comment]
This is a bit of an about-face from my previous statements about functional analyzers (FAs), but my work with the APL functional analyzer has generally convinced me that FAs are a technological dead end for WoWA as a project. I had like 2k words written about the technical journey that led me to this conclusion, but the short version is:

1. Working with the APL FA after initial release ended up being a huge pain in the ass. Reconstructing state from in-stream fabricated events is tedious and error-prone, and numerous bugs that were fixed in `SpellUsable` ages ago had to be re-fixed in the APL code.
1. As I iterated on the FA api to produce something with better DX, I ended on stateful coroutines. This sidesteps many of the data structures problems that would need to be solved for proper pure FAs, while preserving the largely functional style. These would look like:
    ```tsx
    function* analyzer() {
      // setup initial state
      while({ event, ...dependencies } = yield state) {
        // process event
      }
      // produce final result
    }
    ```
2. As I refined this, I came to realize that **our current `Analyzer` pattern is an implementation of stateful coroutines.** This means that the end-state I was exploring with coroutines is just our current status quo with a different coat of paint.
3. The other path (truly pure FAs) is not viable IMHO because of the number of complex data structures that we'd need (range b-trees, immutable data structures w/ state sharing), and which often defeat JS VM optimizations due to their use patterns (unfortunately). Nevermind the fact that we'd need to throw out basically our entire existing analyzer suite to port to them.

In the end, the core problems with our current API are not an "objects vs functions" problem. We have functions with state. Objects are fine for that. Our problems are:

- Constructing analyzers in test is very difficult, but this is largely because we inject too much external state (like the whole CLP!) which requires extensive mocking to construct an `Analyzer` in a test.
- Re-using analysis is often difficult. This frequently happens (IMO) because we put both the analysis and presentation code in the same class. The `MajorDefensives` core analyzer that I added has broken this pattern by exposing an API that is used by multiple different presentation components, and I think that is generally the path forward.
- Our dependency injection API has pretty bad DX (lots of repetition), leading to a lot of boilerplate to add a new Analyzer. But this is tied to the use of `static dependencies` and can be addressed by refactoring (some approaches like the `withDependencies` helper are being explored rn, and [decorator solutions](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#decorators) could be a future path)

As a result, I would like to close this issue and fully deprecate FAs. "Shorthand FAs" that produce classes will continue to exist, but the `statistic` and `suggestion` FA functions will likely go away. Instead, work will continue on improving DX and dealing with architectural debt like the pre-TS DI system.
